File: php/src/detectors/SsrfDetector.php
<?php

namespace FireWTWall\Detectors;

use FireWTWall\Request;

/**
 * Detects Server-Side Request Forgery (SSRF) attempts.
 *
 * Scans query parameters and body fields whose names are commonly used to
 * pass URLs or redirect targets, looking for private IP ranges, cloud metadata
 * endpoints, and dangerous URI schemes.
 *
 * Requires PHP >= 8.0
 */
class SsrfDetector
{
	/** Parameter names that commonly carry a URL or redirect target */
	private const URL_PARAMS = [
		'url', 'redirect', 'return', 'callback', 'next', 'dest', 'destination',
		'src', 'source', 'uri', 'link', 'href', 'proxy', 'forward', 'returnUrl',
		'goto', 'target', 'redir', 'r', 'u',
	];

	/** Patterns that identify private / loopback IP addresses */
	private const PRIVATE_IP_PATTERNS = [
		'/^127\./i',
		'/^10\./i',
		'/^172\.(1[6-9]|2[0-9]|3[01])\./i',
		'/^192\.168\./i',
		'/^0\.0\.0\.0/i',
		'/^::1$/i',
	];

	/** Cloud instance metadata endpoints */
	private const METADATA_PATTERNS = [
		'/169\.254\.169\.254/i',
		'/metadata\.google\.internal/i',
		'/metadata\.azure\.com/i',
		'/100\.100\.100\.200/i',
	];

	/** Dangerous URI schemes (excluding http/https/ftp used legitimately) */
	private const SCHEME_PATTERNS = [
		'/^file:\/\//i',
		'/^gopher:\/\//i',
		'/^dict:\/\//i',
		'/^ldap:\/\//i',
		'/^tftp:\/\//i',
		'/^ftp:\/\//i',
	];

	/**
	 * Scan the request for SSRF indicators.
	 *
	 * @param  Request $request  Normalised request object
	 * @param  array   $config   WAF configuration (reserved for future use)
	 * @return array{rule:string,matched:string,source:string,severity:string}|null
	 */
	public static function scan(Request $request, array $config): ?array
	{
		$allSources = [
			'query' => $request->getQuery(),
			'body'  => $request->getBody(),
		];

		foreach ($allSources as $sourceLabel => $data) {
			$result = self::scanFlat($data, $sourceLabel);
			if ($result !== null) {
				return $result;
			}
		}

		return null;
	}

	// ------------------------------------------------------------------ //
	// Private helpers
	// ------------------------------------------------------------------ //

	/**
	 * Walk a (possibly nested) array and check every key that is a URL param
	 * name for suspicious values.
	 */
	private static function scanFlat(array $data, string $sourceLabel, string $prefix = ''): ?array
	{
		foreach ($data as $key => $value) {
			$paramName = $prefix !== '' ? $prefix . '[' . $key . ']' : (string) $key;

			if (is_array($value)) {
				$result = self::scanFlat($value, $sourceLabel, $paramName);
				if ($result !== null) {
					return $result;
				}
				continue;
			}

			if (!is_string($value)) {
				continue;
			}

			// Only inspect parameters whose name suggests a URL or redirect
			if (!in_array(strtolower((string) $key), array_map('strtolower', self::URL_PARAMS), true)) {
				continue;
			}

			if (self::isSuspiciousValue($value)) {
				return [
					'rule'     => 'ssrf-private-ip',
					'matched'  => $value,
					'source'   => $sourceLabel,
					'severity' => 'critical',
				];
			}
		}

		return null;
	}

	/** Return true if $value matches any SSRF indicator. */
	private static function isSuspiciousValue(string $value): bool
	{
		// Extract host portion for IP checks
		$host = self::extractHost($value);

		foreach (self::PRIVATE_IP_PATTERNS as $pattern) {
			if ($host !== '' && preg_match($pattern, $host)) {
				return true;
			}
		}

		foreach (self::METADATA_PATTERNS as $pattern) {
			if (preg_match($pattern, $value)) {
				return true;
			}
		}

		foreach (self::SCHEME_PATTERNS as $pattern) {
			if (preg_match($pattern, ltrim($value))) {
				return true;
			}
		}

		return false;
	}

	/** Extract the host (or bare value) from a URL-like string. */
	private static function extractHost(string $value): string
	{
		$parsed = @parse_url($value);
		if ($parsed !== false && isset($parsed['host'])) {
			return $parsed['host'];
		}
		// Bare IP/hostname without scheme
		return strtok($value, "/?#:") ?: $value;
	}
}

File: aspnet/src/detectors/SsrfDetector.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    /// <summary>
    /// Detects Server-Side Request Forgery (SSRF) attempts.
    /// </summary>
    public static class SsrfDetector
    {
        private static readonly HashSet<string> UrlParams = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "url", "redirect", "return", "callback", "next", "dest", "destination",
            "src", "source", "uri", "link", "href", "proxy", "forward", "returnurl",
            "goto", "target", "redir", "r", "u",
        };

        private static readonly Regex[] PrivateIpPatterns =
        {
            new Regex(@"^127\.",        RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^10\.",         RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^172\.(1[6-9]|2[0-9]|3[01])\.", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^192\.168\.",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^0\.0\.0\.0",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^::1$",         RegexOptions.IgnoreCase | RegexOptions.Compiled),
        };

        private static readonly Regex[] MetadataPatterns =
        {
            new Regex(@"169\.254\.169\.254",     RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"metadata\.google\.internal", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"metadata\.azure\.com",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"100\.100\.100\.200",      RegexOptions.IgnoreCase | RegexOptions.Compiled),
        };

        private static readonly Regex[] SchemePatterns =
        {
            new Regex(@"^file://",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^gopher://", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^dict://",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^ldap://",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^tftp://",   RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"^ftp://",    RegexOptions.IgnoreCase | RegexOptions.Compiled),
        };

        public static DetectorResult Scan(WafRequest request)
        {
            foreach (var source in new[] {
                ("query", request.Query),
                ("body",  request.Form)
            })
            {
                foreach (var kv in source.Item2)
                {
                    if (!UrlParams.Contains(kv.Key)) continue;
                    if (IsSuspicious(kv.Value))
                        return new DetectorResult("ssrf-private-ip", "critical",
                            kv.Value.Length > 120 ? kv.Value.Substring(0, 120) : kv.Value,
                            source.Item1);
                }
            }
            return null;
        }

        private static bool IsSuspicious(string value)
        {
            string host = ExtractHost(value);

            foreach (var p in PrivateIpPatterns)
                if (host.Length > 0 && p.IsMatch(host)) return true;

            foreach (var p in MetadataPatterns)
                if (p.IsMatch(value)) return true;

            foreach (var p in SchemePatterns)
                if (p.IsMatch(value.TrimStart())) return true;

            return false;
        }

        private static string ExtractHost(string value)
        {
            try
            {
                var uri = new Uri(value);
                return uri.Host;
            }
            catch
            {
                int end = value.IndexOfAny(new[] { '/', '?', '#', ':', '\0' });
                return end >= 0 ? value.Substring(0, end) : value;
            }
        }
    }
}
